### DISCUSSION TITLE
Has anybody looked at adding cellular support to Tasmota

### DISCUSSION THREAD

**Discussion by ajlennon** [arendst/Tasmota#16963] [Ideas]
Hi all,

I'm interested in adding cellular support to Tasmota, say at it's most basic an MQTT connection to integrate into the core Tasmota functionality.

I see a little on cellular when I google but mainly people asking questions not on anything that has been done.

Just wondered if anybody else has done this / would like to talk to me about it as I plan to do some work on this.

Thanks!

Alex

**Comment by sfromis** [comment]
The general suggestion to use those boards has been to use the AT commands via serial to do the communication, likely using ESP32 and the Berry language. I'm not aware of any effort to facilitate MQTT connections via those boards.

**Comment by ajlennon** [comment]
Berry. That is *very* interesting. I hadn't thought of that. Great suggestion thanks!

**Comment by Jason2866** [comment]
Keep in mind Tasmota is not designed to connect directly to the Internet.

**Comment by ajlennon** [comment]
>Keep in mind Tasmota is not designed to connect directly to the Internet.

Thanks! What issues is this likely to cause would you say?

**Reply by barbudor** [reply]
Easy Hacking
There is no firewall in tasmota

**Comment by jkwim** [comment]
You can use it ofcourse to send SMS using AT Commands. I am also thinking of doing it for some alarm messages.

**Reply by ajlennon** [reply]
I very much like this idea. I am thinking of using AT commands to send MQTT messages

**Reply by barbudor** [reply]
I don't think there is any support for MQTT in AT commands
Best you can do is http(s) get/post

**Reply by ajlennon** [reply]
Yes there are various modem modules we've worked with that have an internal TCP/IP stack and support MQTT

e.g. https://www.quectel.com/wp-content/uploads/2021/03/Quectel_BG95BG77BG600L_Series_MQTT_Application_Note_V1.1-1.pdf

**Reply by ajlennon** [reply]
Going back to the point above about securing the device I think limiting support to MQTT message passthrough is quite a nice way to limit the attack surface

**Reply by sfromis** [reply]
Arguably, it *increases* the attack surface, if/when the remote device being allowed general MQTT is not effectively secured.

**Reply by ajlennon** [reply]
Ah yes - I meant in comparison to me integrating full TCP/IP support...

**Comment by ajlennon** [comment]
I thought you might be interested to see some progress I made with a little Berry script to send MQTT messages over a serial connected cellular modem

https://www.youtube.com/watch?v=ZUtPN9FiUeI

**Reply by barbudor** [reply]
Cool

On my side I have started looking at TinyGSM but haven't progress much

**Comment by ajlennon** [comment]
@barbudor I've had a go at bringing in TinyGSM now and it actually comes in quite nicely I think.

https://www.youtube.com/watch?v=9FK0vJz1wrU

**Reply by dimentiy89** [reply]
did you manage to combine the library TinyGSM with tasmota to control via iot?

**Reply by ajlennon** [reply]
I did actually. Still working on it now but I managed to get some life out of it :) It's a bit messy and I was wondering if it's possibly to tidy it up so that I could provide a PR

**Reply by ajlennon** [reply]
I'll let you know when I put a commit into my fork if you like?

**Reply by Cloolalang** [reply]
Hi Alex, I wonder if you got any further with your Tasmota/cellular experiments? I have a Quectel BG96 (LTE Cat M1/NBOIT/GPRS modem and would be really interested to see your Berry script which parses the Tasmota serial to the AT-command MQTT messages for the modem and the decoding in the reverse direction. I, like you would really like to see Tasmota have an alternative method of WAN access.. "Tasmota-mobile". Also I see that the BG96 has on-board GPS, it would be great to be able to get Berry to access and process this location data.  Thanks for posting your youtube video! 
Cheers. Patrick

**Reply by ajlennon** [reply]
H there. I didn't have a lot of success with Berry but I did manage to get TinyGSM in there and get some MQTT communications going.

You can take a look at the fork & branch here. I'm on the Quectel BG600L so it should work for you...

https://github.com/DynamicDevices/Tasmota/tree/ajl/cellular-dev
